base_layer/wallet/src/base_node_service/monitor.rs

```diff
@@ -176,14 +181,34 @@ impl<T: WalletBackend + 'static> BaseNodeMonitor<T> {
         }
     }
 
-    async fn attempt_dial(&mut self, peer: NodeId) -> Result<PeerConnection, BaseNodeMonitorError> {
-        let conn = self.connectivity_manager.dial_peer(peer).await?;
-        Ok(conn)
+    // Establish a peer connection; only redial if the peer changed or is not connected
+    async fn make_peer_connection(&mut self, peer_node_id: NodeId) -> Result<(), BaseNodeMonitorError> {
+        let make_new_connection = match self.connected_peer.clone() {
+            Some(current_connection) => {
+                peer_node_id.clone() != *current_connection.peer_node_id() || !current_connection.is_connected()
+            },
+            None => true,
+        };
+        if make_new_connection {
+            self.connected_peer = Some(self.connectivity_manager.dial_peer(peer_node_id).await?);
+        };
+        Ok(())
     }
 
-    async fn connect_client(&self, mut conn: PeerConnection) -> Result<BaseNodeWalletRpcClient, BaseNodeMonitorError> {
-        let client = conn.connect_rpc().await?;
-        Ok(client)
+    // Return a new RPC session; close the connection if the RPC session fails
+    async fn get_rpc_session(&self) -> Result<BaseNodeWalletRpcClient, BaseNodeMonitorError> {
+        let connection = self
+            .connected_peer
+            .as_ref()
+            .ok_or(BaseNodeMonitorError::NoPeerConnection)?;
+        let rpc_client: BaseNodeWalletRpcClient = match connection.clone().connect_rpc().await {
+            Ok(client) => client,
+            Err(e) => {
+                connection.clone().disconnect().await?;
```

Comment:
So I think this points towards an issue with the management of the connections in the comms layer. This action here is something that should be done by the ConnectionManager when errors occur on that connection.

If this disconnection fixes the issue then it shows that there was something wrong with the connection that ideally the ConnectionManager should have detected.

I am not a fan of papering over the cracks in the domain layer on issues like this because there is clearly an issue in the management of the connections but we are treating the symptom in various places in the domain layer and thus hiding information that would reveal the comms layer issue.

@sdbondi @mikethetike thoughts?

Reply:
If a connection has been "rudely" dropped by the remote peer, it is expected that there will be a delay, typically of a few seconds, before the peer connection recognises that the peer is gone. This is the nature of connections, one has to send something and not receive a TCP ACK, and then there's TOR which may delay this further. We accept that there is a large cost to pinging constantly to check every few milliseconds. So software must be robust to the delay. 

Absolutely agree, that upon trying to establish a rpc session, if the connection is no longer valid, the peer connection should be in a disconnected state. I would expect this over TCP. I have no reason to suspect that this isn't the case for TCP, but we don't exactly have a test where we hard kill a process to test this. However, I suspect that this could be a more complex problem when routed over tor - if you think about it, the final hop of the tor node would have to pass back the information that the destination peer has gone - we don't know how well tor handles this.

So what options do we have, assuming this assessment is accurate:
1. comms should disconnect a peer connection whenever an RPC timeout occurs
     - counter: timeouts are not _always_ indicative of a dropped connection - so disconnecting is likely unexpected behaviour
2. Additional robustness to this edge-case, the wallet should handle a timeout by requesting a new peer connection - if it times out again, then request a new peer connection and so on. If there is a way to reduce the delay over tor, this solution should live in comms but the wallet's behaviour stays as described in this point regardless.

Thanks @hansieodendaal for bringing this to light. As you know, I am working on this at the moment.

Reply:
I understand that the base node may not have been rudely disconnected - if you want this to go in that is perfectly ok with me.

Reply:
To be clear, in this case, comms connectivity reports that the connection has not been dropped by the peer and is still valid. 

Each call in the wallet monitor to process base chain metadata starts with a dial peer to obtain the peer connection, then requesting an RPC client session, then communicating with the RPC client:
``` rust
1.)  let peer = self.wait_for_peer_to_be_set().await?;
2.)  let conn = self.connectivity_manager.dial_peer(peer).await?;       -> PeerConnection
3.)  let client = conn.connect_rpc().await?;                            -> BaseNodeWalletRpcClient
4.)  self.monitor_node(peer, client).await?;
```
When the RPC is misbehaving as described by this PR, (2.) would always succeed en (3.) would always fail, for hours on end. This boils down to `PeerConnection::is_connected()` returning `true` when an RPC client session cannot be established.

I agree with your point 3. above ("the wallet should handle a timeout by requesting a new peer connection"), as it is exactly what this PR proposes. In this case, there is only one peer in the list, so requesting a new peer connection entails disconnecting the current connection, which will force a new connection next time round.

Reply:
> When the RPC is misbehaving as described by this PR, (2.) would always succeed en (3.) would always fail, for hours on end. 

Reproducing this in a test and/or isolating the cause is useful. 

> so requesting a new peer connection entails disconnecting the current connection

If the connection is still connected, this is invalid, there are many reasons a connect_rpc could return an error including the remote peer being too busy. Disconnecting the connection used by other components is not correct. Otherwise, if the connection thinks it's working for hours on end but isn't, the cause should be found and fixed.